### PR TITLE
[READY]Lollipops, gumballs, and a cyborg module that shoots them out!

### PR DESCRIPTION
**Pull request by silicons**
Gives medical cyborgs and service cyborgs lollipop guns.
Has 3 modes
1. dispense lollipops
2. throw lollipops
3. shoot gumballs

Lollipops and gumballs do 5 damage on throw if the cyborg is emagged and can be shot a bit faster then normal projectiles for a bonus.
And the best part? Both have RANDOMIZED COLORS!!!
Spritesand idea from @TehZombehz 

Updates cyborg hug module. :^)

:cl: Mekhi
rscadd: Medical and Service cyborgs now have LOLLIPOP AND GUMBALL SYNTHESIZERS.
experimental: Both of the sugary treats have randomized colors.
experimental: Intensifies the cyborg hugbox modules.
\:cl:

90% sure this does not have any bugs as I'm pretty sure I fixed all of them.....

### CODE REVIEW

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 1]
```diff
@@ -338,4 +338,29 @@
...
+	..(A)
+	throw_speed = 1
+	throwforce = 0
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball
+	name = "gumball"
+	desc = "A colorful, sugary gumball."
+	icon = 'icons/obj/lollipop.dmi'
+	icon_state = "gumball"
+	list_reagents = list("sugar" = 5, "bicardine" = 2, "kelotene" = 2)	//Kek
```
Comment (RandomMarine) [inline code/modules/food_and_drinks/food/snacks_other.dm]: kelotane*

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 2]
```diff
@@ -338,4 +338,29 @@
...
+	..(A)
+	throw_speed = 1
+	throwforce = 0
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball
+	name = "gumball"
+	desc = "A colorful, sugary gumball."
+	icon = 'icons/obj/lollipop.dmi'
+	icon_state = "gumball"
+	list_reagents = list("sugar" = 5, "bicardine" = 2, "kelotane" = 2)	//Kek
```
Comment (RandomMarine) [inline code/modules/food_and_drinks/food/snacks_other.dm]: Forgot the bicaridine too.

**Comment by PKPenguin321** [comment]
Should be added to the cookie dispenser module

**Comment by kevinz000** [comment]
>giving pkborgs all teh fun

**Comment by FantasticFwoosh** [comment]
50/50 not sure what to think, on one hand its pretty novel.

> Module bloat on silly meme that do _terrible_ damage when instead the med (has poison pricks) & the service has stuff like trayspam & guitars etc depending on the specific type.

Not to mention the missed opportunity for a gum into -> rubber plant for hydro

**Comment by kevinz000** [comment]
thanks @FantasticFwoosh

**Comment by kevinz000** [comment]
i dont like botany, it's primary purpose isn't shooting at people.

**Comment by FantasticFwoosh** [comment]
@kevinz000 unless you count gatfruit, poison laced eggs & explosive lemons to be relevant.

**Comment by TehZombehz** [comment]
>Balance discussion around simple children's candy

**Comment by kevinz000** [comment]
Yes.
Also @FantasticFwoosh I meant the borg module, botany has their fun already . If you want to add it to botany, feel free, I already made the code for the foods themselves and the randomized color for you man. It's just that this PR is focused on adding them to medical and service cyborgs, modules that don't have enough love.
*honks*

**Comment by Cebasiz** [comment]
give it to peacekeeper and make emag allow full auto

wew

**Comment by Cebasiz** [comment]
also medical is actually the best module in the game my man so don't add more things it doesn't need

add something so it can help people with blood and remove service cyborg from the game instead thx

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 3]
```diff
@@ -338,4 +338,29 @@
...
 
-
+/obj/item/weapon/reagent_containers/food/snacks/lollipop
+	name = "lollipop"
+	desc = "A delicious lollipop. Makes for a great Valentine's present."
+	icon = 'icons/obj/lollipop.dmi'
+	icon_state = "lollipop_stick"
+	list_reagents = list("nutriment" = 1, "vitamin" = 1, "iron" = 10, "sugar" = 5, "omnizine" = 2)	//Honk
+
+/obj/item/weapon/reagent_containers/food/snacks/lollipop/New()
```
Comment (ChangelingRain) [inline code/modules/food_and_drinks/food/snacks_other.dm]: you need to call ..() in New() procs

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 4]
```diff
@@ -338,4 +338,29 @@
...
+	throwforce = 0
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball
+	name = "gumball"
+	desc = "A colorful, sugary gumball."
+	icon = 'icons/obj/lollipop.dmi'
+	icon_state = "gumball"
+	list_reagents = list("sugar" = 5, "bicaridine" = 2, "kelotane" = 2)	//Kek
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball/New()
```
Comment (ChangelingRain) [inline code/modules/food_and_drinks/food/snacks_other.dm]: same here

**Comment by jughu** [comment]
Only if walking on gumbals make you tril

**Comment by kevinz000** [comment]
@Cebasiz Engiborg is.  :^)

**Comment by jughu** [comment]
Tfw kevinz doesnt asd a emagged function for gumballs where you mass dispense gumballs all over the floor which will slip(trip) people

**Comment by TehZombehz** [comment]
One thing at a time, jughu. Soon(TM)

File: code/game/objects/items/robot/robot_items.dm [hunk 5]
```diff
@@ -326,6 +338,164 @@
...
+
+/obj/item/borg/lollipop/proc/charge_lollipops()
+	candy++
+	charging = 0
+	check_amount()
+
+/obj/item/borg/lollipop/proc/dispense(atom/A, mob/user)
+	if(!candy)
+		user << "<span class='warning'>No lollipops left in storage!</span>"
+		return 0
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: use the DEFINES PLEASE I BEG OF YOU, PLEASE PLEASE PLEASE SIR I BEG OF YOU

File: code/game/objects/items/robot/robot_items.dm [hunk 6]
```diff
@@ -326,6 +338,164 @@
...
+	new /obj/item/weapon/reagent_containers/food/snacks/lollipop(T)
+	candy--
+	check_amount()
+	user << "<span class='notice'>Dispensing lollipop...</span>"
+	playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+	return 1
+
+/obj/item/borg/lollipop/proc/shootL(atom/target, mob/living/user, params)
+	if(!candy)
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: what if get's set to less than zero, use <= 0 instead

File: code/game/objects/items/robot/robot_items.dm [hunk 7]
```diff
@@ -326,6 +338,164 @@
...
+	var/obj/item/ammo_casing/caseless/lollipop/A = new /obj/item/ammo_casing/caseless/lollipop(src)
+	A.BB.damage = hitdamage
+	A.BB.speed = 0.5
+	playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+	A.fire_casing(target, user, params, 0, 0, null, 0)
+	user.visible_message("<span class='warning'>[user] blasts a flying lollipop at [target]!</span>")
+	check_amount()
+
+/obj/item/borg/lollipop/proc/shootG(atom/target, mob/living/user, params)	//Most certainly a good idea.
+	if(!candy)
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: same again

File: code/game/objects/items/robot/robot_items.dm [hunk 8]
```diff
@@ -326,6 +338,164 @@
...
+/obj/item/borg/lollipop/afterattack(atom/target, mob/living/user, proximity, click_params)
+	check_amount()
+	if(iscyborg(user))
+		var/mob/living/silicon/robot/R = user
+		if(R.emagged)
+			hitdamage = emaggedhitdamage
+	switch(mode)
+		if(1)
+			if(!proximity)
+				return 0
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: DEFINES

File: code/game/objects/items/robot/robot_items.dm [hunk 9]
```diff
@@ -326,6 +338,164 @@
...
+		if(3)
+			mode = 1
+			user << "<span class='notice'>Module is now dispensing lollipops.</span>"
+	..()
+
+/obj/item/ammo_casing/caseless/gumball
+	name = "Gumball"
+	desc = "Why are you seeing this?!"
+	projectile_type = /obj/item/projectile/bullet/reusable/gumball
+	click_cooldown_override = 2
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: why have you put this here, please make a comment explaining

File: code/game/objects/items/robot/robot_items.dm [hunk 10]
```diff
@@ -326,6 +338,164 @@
...
+
+
+/obj/item/projectile/bullet/reusable/gumball
+	name = "gumball"
+	desc = "Oh noes! A fast-moving gumball!"
+	icon_state = "gumball"
+	ammo_type = /obj/item/weapon/reagent_containers/food/snacks/gumball
+
+/obj/item/projectile/bullet/reusable/gumball/handle_drop()
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: remember what happened last time you made reusable bullets?

please test this

**Review by optimumtact — DISMISSED** [review]
Needs work

File: code/game/objects/items/robot/robot_items.dm (hunk 10 above)
Comment (kevinz000) [inline code/game/objects/items/robot/robot_items.dm]: i did it works

File: code/game/objects/items/robot/robot_items.dm (hunk 9 above)
Comment (kevinz000) [inline code/game/objects/items/robot/robot_items.dm]: for rapid firing of candy everywhere as normal click delay is too slow? :)

File: code/game/objects/items/robot/robot_items.dm (hunk 10 above)
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: :worried:

**Comment by kevinz000** [comment]
@optimumtact  should be fixed
damage lowered to 3 per shot. shots now take 20 energy regardless.

File: code/game/objects/items/robot/robot_items.dm [hunk 11]
```diff
@@ -371,7 +371,7 @@
 	check_amount()
 
 /obj/item/borg/lollipop/proc/dispense(atom/A, mob/user)
-	if(!candy)
+	if(candy >= 0)
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: if candy is greater than or equal to zero, tell the user there are no lollipops left in storage ??

File: code/game/objects/items/robot/robot_items.dm [hunk 12]
```diff
@@ -371,7 +371,7 @@
 	check_amount()
 
 /obj/item/borg/lollipop/proc/dispense(atom/A, mob/user)
-	if(!candy)
+	if(candy >= 0)
 		user << "<span class='warning'>No lollipops left in storage!</span>"
 		return 0
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: PLEASE I BEG OF YOU

File: code/game/objects/items/robot/robot_items.dm [hunk 13]
```diff
@@ -423,6 +423,9 @@
 	check_amount()
 	if(iscyborg(user))
 		var/mob/living/silicon/robot/R = user
+		if(!R.cell.use(20))
+			user << "<span class='warning'>Not enough power.</span>"
+			return 0
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: AHHHHHHHHHHHH

File: code/game/objects/items/robot/robot_items.dm [hunk 14]
```diff
@@ -326,6 +338,167 @@
...
+		addtimer(src, "charge_lollipops", charge_delay, TIMER_NORMAL)
+		charging = 1
+
+/obj/item/borg/lollipop/proc/charge_lollipops()
+	candy++
+	charging = 0
+	check_amount()
+
+/obj/item/borg/lollipop/proc/dispense(atom/A, mob/user)
+	if(candy >= 0)
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: you want less than or equal to zero, not greater than or equal to zero

File: code/game/objects/items/robot/robot_items.dm [hunk 15]
```diff
@@ -326,6 +338,167 @@
...
+		user << "<span class='warning'>No lollipops left in storage!</span>"
+		return 0
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: PLS

File: code/game/objects/items/robot/robot_items.dm (hunk 14 above)
Comment (kevinz000) [inline code/game/objects/items/robot/robot_items.dm]: ....fuck

**Comment by optimumtact** [comment]
YOu reverted mso's config changes, don't do that

**Comment by kevinz000** [comment]
wait WHAT HOW

**Comment by kevinz000** [comment]
"Fuck"

**Comment by Jordie0608** [comment]
Config names are still incorrect

**Comment by kevinz000** [comment]
fixed

File: code/game/objects/items/robot/robot_items.dm [hunk 16]
```diff
@@ -326,6 +340,167 @@
...
+	candy++
+	charging = 0
+	check_amount()
+
+/obj/item/borg/lollipop/proc/dispense(atom/A, mob/user)
+	if(candy <= 0)
+		user << "<span class='warning'>No lollipops left in storage!</span>"
+		return FALSE
+	var/turf/T = null
+	if(isturf(A))
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: replace this if check chains with

     var/turf/T =  get_turf(A)
     if(!T || !istype(T))
        return FALSE
     new /obj/item/weapon/reagent_containers/food/snacks/lollipop(T)
    
It should work exactly the same and be much cleaner

**Review by optimumtact — CHANGES_REQUESTED** [review]
some final changes that can be made

**Comment by kevinz000** [comment]
Added delayed deletion to delete the shot ones unless someone picks it up as they can be shot extremely quick, and I don't want to be responsible for gumball station 13.

File: code/game/objects/items/robot/robot_items.dm [hunk 17]
```diff
@@ -326,6 +340,159 @@
...
+
+/obj/item/borg/lollipop/dropped()
+	check_amount()
+
+/obj/item/borg/lollipop/proc/check_amount()	//Doesn't even use processing ticks.
+	if(charging)
+		return
+	if(candy < candymax)
+		addtimer(src, "charge_lollipops", charge_delay, TIMER_NORMAL)
+		charging = 1
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: TRUE FALSE DEFINES PLEASE

Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: PLEASE WHY DO I HAVE TO KEEP ASKING I BEG OF YOU KEVIN

File: code/game/objects/items/robot/robot_items.dm [hunk 18]
```diff
@@ -326,6 +340,159 @@
...
+	candy++
+	charging = 0
+	check_amount()
+
+/obj/item/borg/lollipop/proc/dispense(atom/A, mob/user)
+	if(candy <= 0)
+		user << "<span class='warning'>No lollipops left in storage!</span>"
+		return FALSE
+	var/turf/T = get_turf(A)
+	if(!T || !istype(T) || !isopenturf(T))
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: why are you checking for openturfs specifically?

File: code/game/objects/items/robot/robot_items.dm (hunk 8 above)
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: PLEASE I BEG OF YOU

Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: return FALSE

File: code/game/objects/items/robot/robot_items.dm [hunk 19]
```diff
@@ -326,6 +340,159 @@
...
+	desc = "Oh noes! A fast-moving gumball!"
+	icon_state = "gumball"
+	ammo_type = /obj/item/weapon/reagent_containers/food/snacks/gumball/cyborg
+
+/obj/item/projectile/bullet/reusable/gumball/handle_drop()
+	if(!dropped)
+		var/turf/T = get_turf(src)
+		var/obj/item/weapon/reagent_containers/food/snacks/gumball/S = new ammo_type(T)
+		S.color = color
+		dropped = 1
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: dropped = TRUE

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 20]
```diff
@@ -338,4 +338,69 @@
...
+	head.color = C
+	add_overlay(head)
+
+/obj/item/weapon/reagent_containers/food/snacks/lollipop/throw_impact(atom/A)
+	..(A)
+	throw_speed = 1
+	throwforce = 0
+
+/obj/item/weapon/reagent_containers/food/snacks/lollipop/cyborg
+	var/spamchecking = 1
```
Comment (optimumtact) [inline code/modules/food_and_drinks/food/snacks_other.dm]: spamchecking = TRUE

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 21]
```diff
@@ -338,4 +338,69 @@
...
+
+/obj/item/weapon/reagent_containers/food/snacks/lollipop/cyborg/New()
+	..()
+	addtimer(src, "spamcheck", 1200)
+
+/obj/item/weapon/reagent_containers/food/snacks/lollipop/cyborg/equipped(mob/living/user, slot)
+	..(user, slot)
+	spamchecking = 0
```
Comment (optimumtact) [inline code/modules/food_and_drinks/food/snacks_other.dm]: spamchecking = FALSE

File: code/modules/food_and_drinks/food/snacks_other.dm (hunk 4 above)
Comment (optimumtact) [inline code/modules/food_and_drinks/food/snacks_other.dm]: @ChangelingRain what is thsi about?

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 22]
```diff
@@ -338,4 +338,69 @@
...
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball/cyborg
+	var/spamchecking = 1
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball/cyborg/New()
+	..()
+	addtimer(src, "spamcheck", 1200)
+
+/obj/item/weapon/reagent_containers/food/snacks/gumball/cyborg/equipped(mob/living/user, slot)
+	..(user, slot)
```
Comment (optimumtact) [inline code/modules/food_and_drinks/food/snacks_other.dm]: it's good practice to do . = ..(user, slot) 

even if you know the return value isn't used, it makes it easier for future developers if they refactor the proc

Comment (kevinz000) [inline code/modules/food_and_drinks/food/snacks_other.dm]: K

File: code/game/objects/items/robot/robot_items.dm [hunk 23]
```diff
@@ -78,16 +84,21 @@
 					if(M.resting)
 						M.resting = 0
 						M.update_canmove()
+					playsound(loc, 'sound/weapons/thudswoosh.ogg', 50, 1, -1)
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_items.dm]: why type this twice?
it's played no matter if the if or else is checked, just unindent it and leave it at the bottom like it was before

**Comment by kevinz000** [comment]
aaannnnd fixed :^)

**Review by optimumtact — APPROVED** [review]

File: code/game/objects/items/robot/robot_items.dm (hunk 18 above)
Comment (kevinz000) [inline code/game/objects/items/robot/robot_items.dm]: So people can't inject lollipops into walls?

Comment (kevinz000) [inline code/game/objects/items/robot/robot_items.dm]: because otherwise you end up with lollipops inside walls?

**Pull request merged**